### PR TITLE
Add k8s integration

### PR DESCRIPTION
**Pull request by rogaldh**
MEDIA-1218

### CODE REVIEW

File: scripts/config-download.sh [hunk 1]
```diff
@@ -0,0 +1,8 @@
+#!/bin/bash -e
+GITHUB_FILE_PATH="${GITHUB_REPO_PATH}/apps/${GKE_APP_NAME}/ns/${GC_NAMESPACE}/${GKE_APP_NAME}-config.yaml"
+GITHUB_FILE="https://api.github.com/repos/${GITHUB_REPO_OWNER}/${GITHUB_REPO}/contents/${GITHUB_FILE_PATH}"
+
+curl --header "Authorization: token ${GITHUB_API_TOKEN}" \
```
[the remark this answers is not in the document]
Comment (ivanovaleksey) [reply scripts/config-download.sh]: **doesn't** exist

File: scripts/k8s.sh [hunk 2]
```diff
@@ -0,0 +1,8 @@
+#!/bin/bash -e
+source $HOME/google-cloud-sdk/path.bash.inc
+gcloud components install kubectl -q
+gcloud container clusters get-credentials ${GC_CLUSTER_NAME} --zone ${GC_COMPUTE_ZONE} --project ${GC_PROJECT}
```
[the remark this answers is not in the document]
Comment (rogaldh) [reply scripts/k8s.sh]: Agreed but `CLUSTER_ZONE` & `CLUSTER_PROJECT` maybe? Env vars are sorted by name. Same prefix seems adequate.

File: scripts/k8s.sh [hunk 3]
```diff
@@ -0,0 +1,8 @@
...
+kubectl config set-context $(kubectl config current-context) --namespace=${GC_NAMESPACE}
```
[the remark this answers is not in the document]
Comment (rogaldh) [reply scripts/k8s.sh]: Context is defined after credentials request. I'll change that to the explicit ns definition at `kubectl`.

File: scripts/k8s.sh [hunk 4]
```diff
@@ -0,0 +1,8 @@
...
+# define affected namespace for context
+kubectl apply -f ${TRAVIS_BUILD_DIR}/k8s
+kubectl patch deployment ${GKE_APP_NAME} -p "{\"metadata\":{\"annotations\":{\"date\":\"`date +'%s'`\"}}}"
```
[the remark this answers is not in the document]
Comment (rogaldh) [reply scripts/k8s.sh]: Ok

File: scripts/config-download.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (rogaldh) [reply scripts/config-download.sh]: Kind of typo

File: .travis.yml [hunk 5]
```diff
@@ -1,40 +1,60 @@
...
+      before_script: cargo install diesel_cli
+      script: diesel migration run
+    - stage: release
+      rust: stable
       env:
-        - CLOUDSDK_CORE_DISABLE_PROMPTS=1
+        - GC_NAMESPACE=staging GKE_APP_NAME=signals
+      script:
+        - |
+          ./scripts/gcloud.sh
```
[the remark this answers is not in the document]
Comment (rogaldh) [reply .travis.yml]: Nope. Is used twice

File: .travis.yml [hunk 6]
```diff
@@ -1,40 +1,60 @@
 language: rust
-cache: cargo
+cache:
+  directories:
+    - ${HOME}/.cargo
+    - ${TRAVIS_BUILD_DIR}/target
```
[the remark this answers is not in the document]
Comment (rogaldh) [reply .travis.yml]: These are default directories for [`cache: cargo`](https://docs.travis-ci.com/user/caching#Rust-Cargo-cache).

File: .travis.yml [hunk 7]
```diff
@@ -1,40 +1,60 @@
...
+    - ${HOME}/google-cloud-sdk
```
[the remark this answers is not in the document]
Comment (rogaldh) [reply .travis.yml]: This dir is removed unless internal directory is present. That means something weird has happened.

File: .travis.yml [hunk 8]
```diff
@@ -1,40 +1,60 @@
...
 sudo: false
 git:
   depth: 1
 notifications:
   email: false
-matrix:
+jobs:
+  fast_finish: true
```
[the remark this answers is not in the document]
Comment (rogaldh) [reply .travis.yml]: Jobs allow to use serial [stage](https://docs.travis-ci.com/user/build-stages/#What-are-Build-Stages%3F) execution.

File: scripts/k8s.sh (hunk 2 above)
[the remark this answers is not in the document]
Comment (manifest) [reply scripts/k8s.sh]: `ZONE` isn't about cluster

File: .travis.yml (hunk 8 above)
[the remark this answers is not in the document]
Comment (manifest) [reply .travis.yml]: They say it's about failures
https://blog.travis-ci.com/2013-11-27-fast-finishing-builds

File: .travis.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (manifest) [reply .travis.yml]: ok

File: .travis.yml (hunk 8 above)
[the remark this answers is not in the document]
Comment (rogaldh) [reply .travis.yml]: Oh, my bad. It seemed to me you've pointed 12th line.

File: k8s/signals-deployment.yml [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+        - name: cloudsql-proxy
+          image: gcr.io/cloudsql-docker/gce-proxy:1.11
+          command: [
+            "/cloud_sql_proxy"
+          ]
+          args: [
+            "-instances=$(DATABASE_INSTANCE)",
+            "-credential_file=/secrets/cloudsql/credentials.json"
+          ]
+          env:
```
[the remark this answers is not in the document]
Comment (rogaldh) [reply k8s/signals-deployment.yml]: Ok

File: .travis.yml [hunk 10]
```diff
@@ -1,50 +1,75 @@
...
+      script: ./scripts/docker-image-publish.sh signals-rs
+    - stage: migrate
+      rust: stable
+      before_script: cargo install diesel_cli
+      script: diesel migration run
+    - stage: release
+      rust: stable
       env:
-        - CLOUDSDK_CORE_DISABLE_PROMPTS=1
+        - CLUSTER_NAMESPACE=staging CLUSTER_APP=signals
```
Comment (manifest) [inline .travis.yml]: Lets name it `NAMESPACE` and `APP`

File: .travis.yml [hunk 11]
```diff
@@ -1,50 +1,75 @@
...
+      script:
+        - |
+          ./scripts/gcloud-install.sh
+          ./scripts/k8s-install.sh
+          ./scripts/k8s-chart-download.sh ${TRAVIS_BUILD_DIR}/k8s
+          ./scripts/k8s-charts-apply.sh ${TRAVIS_BUILD_DIR}/k8s
```
Comment (manifest) [inline .travis.yml]: `k8s.charts.apply.sh`
We use this naming convention.

File: k8s/signals-config.yml [hunk 12]
```diff
@@ -0,0 +1,8 @@
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: signals-sample
+data:
+  mqtt-host: MQTT_HOST
+  mqtt-client-id: MQTT_CLIENT_ID
+  database-instance: DATABASE_INSTANCE
```
Comment (manifest) [inline k8s/signals-config.yml]: `MQTT_HOST`, `MQTT_CLIENT_ID`, `DATABASE_INSTANCE` ?

File: scripts/gcloud-install.sh [hunk 13]
```diff
@@ -0,0 +1,10 @@
+#!/bin/bash -e
+CRED_FILE=${CRED_FILE:-'.travis-key.json'}
```
Comment (manifest) [inline scripts/gcloud-install.sh]: SERVICE_ACCOUNT_KEYPATH=${SERVICE_ACCOUNT_KEYPATH:-'/tmp/travis-ci-key.json'}

File: scripts/k8s-chart-download.sh [hunk 14]
```diff
@@ -0,0 +1,21 @@
+#!/bin/bash -e
+[[ ! ${CLUSTER_APP} ]] && $(>&2 echo "CLUSTER_APP is not specified") && exit 1
+[[ ! ${CLUSTER_NAME} ]] && $(>&2 echo "CLUSTER_NAME is not specified") && exit 1
+[[ ! ${CLUSTER_NAMESPACE} ]] && $(>&2 echo "CLUSTER_NAMESPACE is not specified") && exit 1
+[[ ! ${GITHUB_API_TOKEN} ]] && $(>&2 echo "GITHUB_API_TOKEN is not specified") && exit 1
+[[ ! ${GITHUB_REPO_OWNER} ]] && $(>&2 echo "GITHUB_REPO_OWNER is not specified") && exit 1
+[[ ! ${GITHUB_REPO_PATH} ]] && $(>&2 echo "GITHUB_REPO_PATH is not specified") && exit 1
+[[ ! ${GITHUB_REPO} ]] && $(>&2 echo "GITHUB_REPO is not specified") && exit 1
+
+FILE_TO_DOWNLOAD=${2:-"${CLUSTER_APP}-config.yaml"}
```
Comment (manifest) [inline scripts/k8s-chart-download.sh]: Better to use named arguments. We already do it.

File: scripts/k8s-chart-download.sh [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+PATH_TO_FILE="${1:-$(pwd)}/${FILE_TO_DOWNLOAD}"
```
Comment (manifest) [inline scripts/k8s-chart-download.sh]: Better to use named arguments

File: scripts/k8s-charts-apply.sh [hunk 16]
```diff
@@ -0,0 +1,8 @@
+#!/bin/bash -e
+[[ ! ${1} ]] && $(>&2 echo "path/to/chart[s] should be defined") && exit 1
```
Comment (manifest) [inline scripts/k8s-charts-apply.sh]: Better to use named arguments

File: scripts/k8s-install.sh [hunk 17]
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash -e
+GCLOUD_SDK_PATH="${HOME}/${GCLOUD_SDK_DIR:-'google-cloud-sdk'}"
+
+[[ ! ${CLUSTER_NAME} ]] && $(>&2 echo "CLUSTER_NAME is not specified") && exit 1
+[[ ! ${CLUSTER_PROJECT} ]] && $(>&2 echo "CLUSTER_PROJECT is not specified") && exit 1
```
Comment (manifest) [inline scripts/k8s-install.sh]: Just `PROJECT`

Comment (manifest) [reply scripts/k8s-install.sh]: Project is bigger then cluster

File: scripts/k8s-chart-download.sh [hunk 18]
```diff
@@ -0,0 +1,21 @@
...
+
+GITHUB_FILE="https://api.github.com/repos/${GITHUB_REPO_OWNER}/${GITHUB_REPO}/contents/${GITHUB_REPO_PATH}/${CLUSTER_NAME}/apps/${CLUSTER_APP}/ns/${CLUSTER_NAMESPACE}/${FILE_TO_DOWNLOAD}"
+
+curl -o ${PATH_TO_FILE} \
+    --header "Authorization: token ${GITHUB_API_TOKEN}" \
+    --header "Accept: application/vnd.github.v3.raw" \
+    --remote-name \
+    --location ${GITHUB_FILE}
+
+cat ${PATH_TO_FILE} | jq -e . >/dev/null 2>&1 && (echo 'File was not found'; exit 1)
```
Comment (manifest) [inline scripts/k8s-chart-download.sh]: `if [[ ! -f ${PATH_TO_FILE} ]]; then $(>&2 echo "PATH_TO_FILE isn't exist"); exit 1; fi`

File: scripts/docker-image-publish.sh [hunk 19]
```diff
@@ -0,0 +1,13 @@
+#!/bin/bash -e
+[[ ! ${DOCKER_IMAGE_OWNER} ]] && $(>&2 echo "DOCKER_IMAGE_OWNER is not specified") && exit 1
+[[ ! ${DOCKER_PASSWORD} ]] && $(>&2 echo "DOCKER_PASSWORD is not specified") && exit 1
+[[ ! ${DOCKER_USERNAME} ]] && $(>&2 echo "DOCKER_USERNAME is not specified") && exit 1
+
+DOCKER_IMAGE_NAME=${1}
```
Comment (manifest) [inline scripts/docker-image-publish.sh]: What the purpose to use `${1}`?

File: k8s/signals-config.yml (hunk 12 above)
Comment (rogaldh) [reply k8s/signals-config.yml]: This meant there should be values. Let's make them empty

File: .travis.yml (hunk 10 above)
Comment (rogaldh) [reply .travis.yml]: Ok

File: .travis.yml (hunk 11 above)
Comment (rogaldh) [reply .travis.yml]: Ok

File: scripts/gcloud-install.sh (hunk 13 above)
Comment (rogaldh) [reply scripts/gcloud-install.sh]: Why not

File: .travis.yml (hunk 11 above)
Comment (ivanovaleksey) [reply .travis.yml]: Isn't strange to name file with multiple dots? Does it give any advantages?

File: scripts/k8s-chart-download.sh (hunk 14 above)
Comment (rogaldh) [reply scripts/k8s-chart-download.sh]: Positional arguments allow to change "what to download" easier compared to named arguments. Also we can use this script multiple times here with the same ENV configuration.

File: scripts/k8s-chart-download.sh (hunk 15 above)
Comment (rogaldh) [reply scripts/k8s-chart-download.sh]: That script has same behaviour as `scripts/k8s-chart-download.sh`.

File: scripts/k8s-charts-apply.sh (hunk 16 above)
Comment (rogaldh) [reply scripts/k8s-charts-apply.sh]: `K8s` directory is project specific. ENV vars might change but project won't. That might come with bugs. Using like so: `./scripts/k8s.charts.apply.sh ${TRAVIS_BUILD_DIR}/k8s` seems more consistent. Positional args plays well here.
Also this check prevent us to apply some unwanted charts unintentionally.

File: scripts/k8s-install.sh (hunk 17 above)
Comment (rogaldh) [reply scripts/k8s-install.sh]: Just wanna give same prefix for platform-specific variables. Let's use PROJECT.

File: scripts/k8s-chart-download.sh (hunk 18 above)
Comment (rogaldh) [reply scripts/k8s-chart-download.sh]: `curl` returns JSON-like response if smth goes wrong with request to github. `.yml` is the superset of JSON and this check seems valid.

File: scripts/docker-image-publish.sh (hunk 19 above)
Comment (rogaldh) [reply scripts/docker-image-publish.sh]: Script becomes reusable: `./scripts/docker.image.publish.sh signals-rs`

File: .travis.yml (hunk 11 above)
Comment (rogaldh) [reply .travis.yml]: I don't think difference between `.` and `-` is crucial. I prefer kebab-case but if team feel well with dots why not?

**Pull request merged**